### PR TITLE
Rework coupon order and invoice record.

### PR DESCRIPTION
**Pull request by mjfrey**
This PR re-works the way we record orders for Coupons and Invoices.

Changes include:
- Create new InvoiceClient model
- Added InvoiceClient to the Django admin
- Add following FKs to Invoice model
  - Order
  - InvoiceClient
- Added Invoices the the Django admin
- Updated Coupon API
  - request.user for order
  - Create Invoice using Order and InvoiceClient
- Update Invoice Payment Processor to user Order and InvoiceClient
- Updated tests

### CODE REVIEW

**Comment by mjfrey** [comment]
@clintonb @vkaracic can you please take a look.

File: ecommerce/invoice/models.py [hunk 1]
```diff
@@ -10,20 +10,22 @@ class Invoice(TimeStampedModel):
...
+    basket = models.ForeignKey('basket.Basket', null=True, blank=False)
+    order = models.ForeignKey('order.Order', null=True, blank=False)
+    invoice_client = models.ForeignKey('core.InvoiceClient', null=True, blank=False)
     state = models.CharField(max_length=255, default=NOT_PAID, choices=state_choices)
 
     history = HistoricalRecords()
 
     def __str__(self):
-        return 'Invoice {id} for order number {order}'.format(id=self.id, order=self.basket.order.number)
+        return 'Invoice {id} for order number {order}'.format(id=self.id, order=self.order)
```
Comment (vkaracic) [inline ecommerce/invoice/models.py]: Should the order part be from `self.order.number` ?

**Comment by vkaracic** [comment]
I presume the idea is that the user who creates the coupon is bound to the basket and the client is to the order now?

**Comment by mjfrey** [comment]
@vkaracic The order is now recorded as who creates the Coupon.  The Invoice contains the client and a link to the order.

File: ecommerce/core/models.py [hunk 2]
```diff
@@ -62,5 +62,14 @@ def get_full_name(self):
 
 
 class Client(User):
-    """The model for the business client."""
     pass
+
+
+class InvoiceClient(models.Model):
```
Comment (clintonb) [inline ecommerce/core/models.py]: `BusinessClient`, or anything not related to invoices. This is more than just invoices.

File: ecommerce/core/models.py [hunk 3]
```diff
@@ -62,5 +62,14 @@ def get_full_name(self):
...
+    """The model for the business client."""
+
+    full_name = models.CharField(_('Full Name'), unique=True, default="",
```
Comment (clintonb) [inline ecommerce/core/models.py]: This field should just be `name`. Remove the default.

File: ecommerce/core/tests/test_models.py [hunk 4]
```diff
@@ -38,3 +38,10 @@ def test_get_full_name(self):
 
         user = self.create_user(full_name=full_name, first_name=first_name, last_name=last_name)
         self.assertEquals(user.get_full_name(), full_name)
+
+
+class InvoiceClientTests(TestCase):
+
+    def test_full_name(self):
```
Comment (clintonb) [inline ecommerce/core/tests/test_models.py]: No need to test basic model fields. Test the `str` method instead.

File: ecommerce/core/models.py [hunk 5]
```diff
@@ -62,5 +62,14 @@ def get_full_name(self):
 
 
 class Client(User):
```
Comment (vkaracic) [inline ecommerce/core/models.py:71]: Can we remove this class?

File: ecommerce/core/tests/test_models.py (hunk 4 above)
Comment (vkaracic) [inline ecommerce/core/tests/test_models.py]: Isn't that what's being tested here?

Comment (clintonb) [inline ecommerce/core/tests/test_models.py]: It is, actually. The test is mis-named.

File: ecommerce/core/models.py (hunk 5 above)
Comment (mjfrey) [inline ecommerce/core/models.py:71]: Not yet.  Need to migrate all data from it first.  Then drop.

**Comment by mjfrey** [comment]
@clintonb @vkaracic Comments addressed.  Can you give another look?

File: ecommerce/core/admin.py [hunk 6]
```diff
@@ -21,5 +21,9 @@ class EcommerceUserAdmin(UserAdmin):
     )
 
 
+@admin.register(BusinessClient)
+class InvoiceClientAdmin(admin.ModelAdmin):
```
Comment (clintonb) [inline ecommerce/core/admin.py]: Correct the name.

File: ecommerce/invoice/models.py [hunk 7]
```diff
@@ -10,20 +10,22 @@ class Invoice(TimeStampedModel):
         (NOT_PAID, _('Not Paid')),
         (PAID, _('Paid')),
     )
-    basket = models.ForeignKey('basket.Basket', null=False, blank=False)
+    basket = models.ForeignKey('basket.Basket', null=True, blank=False)
```
Comment (clintonb) [inline ecommerce/invoice/models.py]: `blank=True`

**Comment by clintonb** [comment]
:+1: once comments addressed.

**Comment by vkaracic** [comment]
Sorry, I forgot to leave the emoticon
:+1:

**Pull request merged**